lib/PrometheusPlugin.js

```diff
@@ -0,0 +1,250 @@
+    });
+
+    this.kuzzleMetrics = {
+      requests: new Prometheus.Histogram({
+        name: `kuzzle_requests`,
+        help: `Kuzzle monitoring succeeded requests`,
+        buckets: Prometheus.exponentialBuckets(0.0001, 1.5, 36),
+        labelNames: ['controller', 'action', 'event', 'status', 'protocol']
+      }),
+      rooms: new Prometheus.Gauge({
+        name: `kuzzle_active_rooms`,
+        help: `Kuzzle active rooms counter`
+      })
+    };
+
+    this.hooks = {
+      'room:new': () => this.recordRooms('room:new'),
+      'room:remove': () => this.recordRooms('room:remove'),
+      'request:onSuccess': this.recordRequests.bind(this),
+      'request:onError': this.recordRequests.bind(this)
+    };
+
+    this.systemMetricsJob = Prometheus.collectDefaultMetrics({
+      register: this.register,
+      timeout: this.config.systemMetricsInterval
+    });
+
+    this.syncJob = setInterval(
+      () => this.pushRegistry(),
+      this.config.syncInterval
+    );
+  }
+
+  /**
+   * Hooks
+   */
+
+  /**
+   * Record rooms creation to Prometheus Registry
+   * when 'room:new' Kuzzle Event is triggered.
+   *
+   * @param {String} event
+   */
+  recordRooms(event) {
+    switch (event) {
+      case 'room:new':
+        this.kuzzleMetrics.rooms.inc();
+        break;
+      case 'room:remove':
+        this.kuzzleMetrics.rooms.dec();
+        break;
+    }
+  }
+
+  /**
+   * Record requests information to Prometheus Registry.
+   *
+   * @param {Request} request
+   * @param {String}  event
+   */
+  recordRequests(request, event) {
+    if (request.input.action !== 'metrics') {
+      this.kuzzleMetrics.requests.observe(
+        {
+          controller: request.input.controller,
+          action: request.input.action,
+          event,
+          protocol: request.context.connection.protocol
+        },
+        Date.now() - request.timestamp
+      );
+    }
+  }
+
+  /**
+   * Route
+   */
+
+  /**
+   * Custom route to fetch all saved metrics
+   * in Prometheus metrics format
+   *
+   * @param {Request} request
+   */
+  async metrics(request) {
+    const register = await this.syncRegisters();
+    request.response.setHeader('Content-Type', register.contentType);
+    request.response.raw = true;
+    return register.metrics();
+  }
+
+  /**
+   * Helpers
+   */
+
+  /**
+   * Push given Prometheus registers as string
+   * to Redis using the provided key
+   */
+  async pushRegistry() {
+    await this.context.accessors.sdk.ms.set(
+      this.key,
+      JSON.stringify(this.registry.getMetricsAsJSON())
+    );
+  }
+
+  /**
+   * Pull Prometheus registers stored in Redis merge them
+   * with the local one and return a new Registry which
+   * contains all Prometheus metrics
+   *
+   * @return {Prometheus.Registry}
+   */
+  async syncRegisters() {
+    let cursor = 0,
+      values = [];
+
+    do {
+      let result = await this.context.accessors.sdk.ms.scan(cursor, {
+        match: 'kuzzle_prometheus_*'
```

Comment:
This will return only the keys found on the queried Redis cluster node. To make sure that you get all results in a cluster, you need to ask Redis to group all these keys in 1 node, by embedding the common key part between brackets.

See https://redis.io/topics/cluster-spec#keys-hash-tags